sourcegraph/sourcegraph
# Better link previews for private instances

Introduces a new `OpenGraphMetadataMiddleware` that serves a separate template with OpenGraph metadata meant for unauthenticated requests to private instances from social bots (e.g. Slackbot). Instead of redirecting the bots to the sign-in page, they can parse the OpenGraph metadata and produce a nicer link preview for a subset of Sourcegraph app routes.

⚠️ The `OpenGraphMetadataMiddleware` does not access any private data on the instance, it purely extracts the data already available in the URL and presents it in a nicer manner.

## Screenshot:

(Ignore the long ugly ngrok URL, imagine `sourcegraph.sourcegraph.com` instead 🙂)

<img width="1178" alt="Screenshot 2022-09-21 at 15 12 29" src="https://user-images.githubusercontent.com/6417322/191513102-f478f4b9-415d-4e39-a4dd-fbf1c404f209.png">

## Test plan


* Run Sourcegraph in enterprise mode
* Run ngrok with `ngrok http --host-header=rewrite 3080` to expose your instance to the world
* Navigate to the ngrok URL (something like `https://c70e-2a00-ee2-4e06-b400-820-34aa-bcc3-1523.eu.ngrok.io`) and copy a repo/blob/search URL
* Paste it into Slack
* You should get a link preview with data from the URL and not the sign-in page
